# Data Analysis on New Jersey School Internet Speeds

In February 2018, New Jersey Department of Education released in its Performance Report 2016-17 data on tech readiness for each school. Raw dataset can be found [here](https://rc.doe.state.nj.us/ReportsDatabase.aspx?).

### [Click here to go straight to the analysis](#analysis)

In [601]:
import pandas as pd

In [602]:
filename = "raw_data/PerformanceReports.xlsx"

In [603]:
performance_report = pd.read_excel(filename, sheetname=None)

In [604]:
performance_report.keys()

[u'GraduationRates',
 u'PostsecondaryEnrRates16mos',
 u'StudentSuspensionRates',
 u'Narrative',
 u'ELALiteracyPerformanceByGrade',
 u'APIBCourseworkPartPerf',
 u'EnrollmentTrendsByGrade',
 u'VisualAndPerformingArts',
 u'PSAT-SAT-ACTPerformance',
 u'DropoutRateTrends',
 u'SchoolHeader',
 u'AlternateAssessmentParticipatio',
 u'ELAPerformanceTrends',
 u'MathCourseParticipation',
 u'StudentGrowth',
 u'Biology',
 u'StudentGrowthByPerformLevel',
 u'TeachersAdminsOneYearRetention',
 u'MathPerformanceTrends',
 u'DaysAbsent',
 u'IndustryValuedCredentialsEarned',
 u'PSAT-SAT-ACTParticipation',
 u'PreKAndK-FullDayHalfDay',
 u'CTE_SLEParticipation',
 u'MathPerformanceByGradeTest',
 u'EnrollmentTrendsByStudentGroup',
 u'ChronicAbsByGrade',
 u'TeachersExperienceCertification',
 u'APIBCoursesOffered',
 u'ELALiteracyParticipationPerform',
 u'WorldLanguagesCourseParticipati',
 u'AccountabilityIndicator',
 u'StudentGrowthByGrade',
 u'GraduationRateTrends',
 u'ScienceCourseParticipation',
 u'EnglishLangu

In [605]:
school_headers = performance_report['SchoolHeader']

In [606]:
school_headers.head(5)

County Code  District Code  School Code County Name  District Name  \
0            1             10           50    ATLANTIC   ABSECON CITY   
1            1             10           60    ATLANTIC   ABSECON CITY   
2            1            110           10    ATLANTIC  ATLANTIC CITY   
3            1            110           30    ATLANTIC  ATLANTIC CITY   
4            1            110           50    ATLANTIC  ATLANTIC CITY   

                 School Name Grade Span  
0             Emma C Attales      05-08  
1             H Ashton Marsh      PK-04  
2  Atlantic City High School      09-12  
3    Sovereign Avenue School      PK-08  
4     Chelsea Heights School      PK-08

In [607]:
tech_readiness = performance_report['TechnologyReadiness']

In [608]:
tech_readiness.head(5)

CountyCode  DistrictCode  SchoolCode StudentDeviceRation InternetSpeed  \
0           1            10          50               1:1.3    120.0 kbps   
1           1            10          60               1:0.4    120.0 kbps   
2           1           110          10               1:1.3     71.4 kbps   
3           1           110          30               1:0.9     71.4 kbps   
4           1           110          50               1:1.3     71.4 kbps   

  RecommendedInternetSpeed MetRecommendedSpeed ConnectivityBetweenSchools  \
0                 100 kbps                 Yes                          N   
1                 100 kbps                 Yes                          N   
2                 100 kbps                  No                          N   
3                 100 kbps                  No                          N   
4                 100 kbps                  No                          N   

  RecommendedConnectivity MetRecommendedConnectivity  
0                   Fiber                         No  
1                   Fiber                         No  
2                   Fiber                         No  
3                   Fiber                         No  
4                   Fiber                         No

In [609]:
len(school_headers.index)

2516

In [610]:
len(tech_readiness.index)

2236

In [611]:
list(school_headers.columns)

[u'County Code',
 u'District Code',
 u'School Code',
 u'County Name',
 u'District Name',
 u'School Name',
 u'Grade Span']

In [612]:
list(tech_readiness.columns)

[u'CountyCode',
 u'DistrictCode',
 u'SchoolCode',
 u'StudentDeviceRation',
 u'InternetSpeed',
 u'RecommendedInternetSpeed',
 u'MetRecommendedSpeed',
 u'ConnectivityBetweenSchools',
 u'RecommendedConnectivity',
 u'MetRecommendedConnectivity']

In [613]:
merge_schools_tech =  pd.merge(school_headers, tech_readiness, left_on=['County Code', 'District Code', 'School Code'], right_on=['CountyCode', 'DistrictCode', 'SchoolCode'])

In [614]:
total_schools = len(merge_schools_tech.index)
len(merge_schools_tech.index)

2236

In [615]:
merge_schools_tech.head(5)

County Code  District Code  School Code County Name  District Name  \
0            1             10           50    ATLANTIC   ABSECON CITY   
1            1             10           60    ATLANTIC   ABSECON CITY   
2            1            110           10    ATLANTIC  ATLANTIC CITY   
3            1            110           30    ATLANTIC  ATLANTIC CITY   
4            1            110           50    ATLANTIC  ATLANTIC CITY   

                 School Name Grade Span  CountyCode  DistrictCode  SchoolCode  \
0             Emma C Attales      05-08           1            10          50   
1             H Ashton Marsh      PK-04           1            10          60   
2  Atlantic City High School      09-12           1           110          10   
3    Sovereign Avenue School      PK-08           1           110          30   
4     Chelsea Heights School      PK-08           1           110          50   

  StudentDeviceRation InternetSpeed RecommendedInternetSpeed  \
0               1:1.3    120.0 kbps                 100 kbps   
1               1:0.4    120.0 kbps                 100 kbps   
2               1:1.3     71.4 kbps                 100 kbps   
3               1:0.9     71.4 kbps                 100 kbps   
4               1:1.3     71.4 kbps                 100 kbps   

  MetRecommendedSpeed ConnectivityBetweenSchools RecommendedConnectivity  \
0                 Yes                          N                   Fiber   
1                 Yes                          N                   Fiber   
2                  No                          N                   Fiber   
3                  No                          N                   Fiber   
4                  No                          N                   Fiber   

  MetRecommendedConnectivity  
0                         No  
1                         No  
2                         No  
3                         No  
4                         No

In [616]:
columns = ['CountyCode', 'DistrictCode', 'SchoolCode']
merge_schools_tech.drop(columns, inplace=True, axis=1)

In [617]:
merge_schools_tech.head(5)

County Code  District Code  School Code County Name  District Name  \
0            1             10           50    ATLANTIC   ABSECON CITY   
1            1             10           60    ATLANTIC   ABSECON CITY   
2            1            110           10    ATLANTIC  ATLANTIC CITY   
3            1            110           30    ATLANTIC  ATLANTIC CITY   
4            1            110           50    ATLANTIC  ATLANTIC CITY   

                 School Name Grade Span StudentDeviceRation InternetSpeed  \
0             Emma C Attales      05-08               1:1.3    120.0 kbps   
1             H Ashton Marsh      PK-04               1:0.4    120.0 kbps   
2  Atlantic City High School      09-12               1:1.3     71.4 kbps   
3    Sovereign Avenue School      PK-08               1:0.9     71.4 kbps   
4     Chelsea Heights School      PK-08               1:1.3     71.4 kbps   

  RecommendedInternetSpeed MetRecommendedSpeed ConnectivityBetweenSchools  \
0                 100 kbps                 Yes                          N   
1                 100 kbps                 Yes                          N   
2                 100 kbps                  No                          N   
3                 100 kbps                  No                          N   
4                 100 kbps                  No                          N   

  RecommendedConnectivity MetRecommendedConnectivity  
0                   Fiber                         No  
1                   Fiber                         No  
2                   Fiber                         No  
3                   Fiber                         No  
4                   Fiber                         No

In [618]:
merge_schools_tech.tail(5)

County Code  District Code  School Code County Name  \
2231           80           8010          980    CHARTERS   
2232           80           8050          990    CHARTERS   
2233           80           8060          990    CHARTERS   
2234           80           8065          980    CHARTERS   
2235           80           8140          990    CHARTERS   

              District Name                             School Name  \
2231  Union County TEAMS CS       Union County TEAMS Charter School   
2232               Unity CS                    Unity Charter School   
2233  University Academy CS  University Academy Charter High School   
2234  University Heights CS       University Heights Charter School   
2235             Village CS              The Village Charter School   

     Grade Span StudentDeviceRation InternetSpeed RecommendedInternetSpeed  \
2231      KG-12               1:0.9    912.6 kbps                 100 kbps   
2232      KG-08               1:0.9    319.1 kbps                 100 kbps   
2233      09-12               1:1.1    238.7 kbps                 100 kbps   
2234      PK-08               1:0.7   1422.5 kbps                 100 kbps   
2235      KG-08               1:1.5    561.8 kbps                 100 kbps   

     MetRecommendedSpeed ConnectivityBetweenSchools RecommendedConnectivity  \
2231                 Yes                          N                   Fiber   
2232                 Yes                          N                   Fiber   
2233                 Yes                          N                   Fiber   
2234                 Yes                          N                   Fiber   
2235                 Yes                      Fiber                   Fiber   

     MetRecommendedConnectivity  
2231                         No  
2232                         No  
2233                         No  
2234                         No  
2235                        Yes

In [687]:
unique_internet_speeds = merge_schools_tech["InternetSpeed"].unique()

In [620]:
merge_schools_tech.dtypes

County Code                    int64
District Code                  int64
School Code                    int64
County Name                   object
District Name                 object
School Name                   object
Grade Span                    object
StudentDeviceRation           object
InternetSpeed                 object
RecommendedInternetSpeed      object
MetRecommendedSpeed           object
ConnectivityBetweenSchools    object
RecommendedConnectivity       object
MetRecommendedConnectivity    object
dtype: object

In [621]:
#How many met recommended Internet Speed

recommended_speed = merge_schools_tech.groupby(["MetRecommendedSpeed"]).count()

recommended_speed = recommended_speed[['School Code']]


In [622]:
recommended_speed.reset_index()

MetRecommendedSpeed  School Code
0                  No          442
1                 Yes         1794

In [623]:
recommended_speed['Percent'] = (recommended_speed['School Code']/total_schools)*100

In [624]:
recommended_speed = recommended_speed.reset_index()

In [625]:
recommended_speed

MetRecommendedSpeed  School Code    Percent
0                  No          442  19.767442
1                 Yes         1794  80.232558

In [626]:
merge_schools_tech["InternetSpeed"] = merge_schools_tech["InternetSpeed"].str.rstrip(' kbps')

In [627]:
merge_schools_tech.dtypes

County Code                    int64
District Code                  int64
School Code                    int64
County Name                   object
District Name                 object
School Name                   object
Grade Span                    object
StudentDeviceRation           object
InternetSpeed                 object
RecommendedInternetSpeed      object
MetRecommendedSpeed           object
ConnectivityBetweenSchools    object
RecommendedConnectivity       object
MetRecommendedConnectivity    object
dtype: object

In [628]:
merge_schools_tech.head(5)

County Code  District Code  School Code County Name  District Name  \
0            1             10           50    ATLANTIC   ABSECON CITY   
1            1             10           60    ATLANTIC   ABSECON CITY   
2            1            110           10    ATLANTIC  ATLANTIC CITY   
3            1            110           30    ATLANTIC  ATLANTIC CITY   
4            1            110           50    ATLANTIC  ATLANTIC CITY   

                 School Name Grade Span StudentDeviceRation InternetSpeed  \
0             Emma C Attales      05-08               1:1.3         120.0   
1             H Ashton Marsh      PK-04               1:0.4         120.0   
2  Atlantic City High School      09-12               1:1.3          71.4   
3    Sovereign Avenue School      PK-08               1:0.9          71.4   
4     Chelsea Heights School      PK-08               1:1.3          71.4   

  RecommendedInternetSpeed MetRecommendedSpeed ConnectivityBetweenSchools  \
0                 100 kbps                 Yes                          N   
1                 100 kbps                 Yes                          N   
2                 100 kbps                  No                          N   
3                 100 kbps                  No                          N   
4                 100 kbps                  No                          N   

  RecommendedConnectivity MetRecommendedConnectivity  
0                   Fiber                         No  
1                   Fiber                         No  
2                   Fiber                         No  
3                   Fiber                         No  
4                   Fiber                         No

In [629]:
merge_schools_tech["InternetSpeed"] = pd.to_numeric(merge_schools_tech["InternetSpeed"], errors='coerce')

In [630]:
merge_schools_tech.dtypes

County Code                     int64
District Code                   int64
School Code                     int64
County Name                    object
District Name                  object
School Name                    object
Grade Span                     object
StudentDeviceRation            object
InternetSpeed                 float64
RecommendedInternetSpeed       object
MetRecommendedSpeed            object
ConnectivityBetweenSchools     object
RecommendedConnectivity        object
MetRecommendedConnectivity     object
dtype: object

In [631]:
merge_schools_tech = merge_schools_tech.sort_values('InternetSpeed')

In [632]:
merge_schools_tech.head(25)

County Code  District Code  School Code County Name  \
275             3           4870           50      BERGEN   
1939           37           1980           50      SUSSEX   
126             3           1470           60      BERGEN   
127             3           1470           80      BERGEN   
2206           80           6665          930    CHARTERS   
1829           31           5690           60     PASSAIC   
1830           31           5690           70     PASSAIC   
2211           80           6915          950    CHARTERS   
109             3           1345           90      BERGEN   
108             3           1345           80      BERGEN   
107             3           1345           70      BERGEN   
106             3           1345           60      BERGEN   
105             3           1345           50      BERGEN   
264             3           4470           50      BERGEN   
208             3           3710           50      BERGEN   
209             3           3710           60      BERGEN   
421             5           5890           30  BURLINGTON   
1788           31           4010          310     PASSAIC   
1752           31           4010           20     PASSAIC   
1787           31           4010          309     PASSAIC   
1789           31           4010          311     PASSAIC   
1784           31           4010          306     PASSAIC   
1786           31           4010          308     PASSAIC   
1754           31           4010           43     PASSAIC   
1783           31           4010          305     PASSAIC   

                               District Name  \
275                     SOUTH HACKENSACK TWP   
1939                             HAMPTON TWP   
126                            FAIRVIEW BORO   
127                            FAIRVIEW BORO   
2206                                 Gray CS   
1829                           WOODLAND PARK   
1830                           WOODLAND PARK   
2211  Jersey City Golden Door Charter School   
109                             ELMWOOD PARK   
108                             ELMWOOD PARK   
107                             ELMWOOD PARK   
106                             ELMWOOD PARK   
105                             ELMWOOD PARK   
264                        ROCHELLE PARK TWP   
208                 NORTHERN VALLEY REGIONAL   
209                 NORTHERN VALLEY REGIONAL   
421                             WOODLAND TWP   
1788                           PATERSON CITY   
1752                           PATERSON CITY   
1787                           PATERSON CITY   
1789                           PATERSON CITY   
1784                           PATERSON CITY   
1786                           PATERSON CITY   
1754                           PATERSON CITY   
1783                           PATERSON CITY   

                                            School Name Grade Span  \
275                                     Memorial School      PK-08   
1939                MARIAN E. MCKEOWN ELEMENTARY SCHOOL      KG-06   
126                                      Lincoln School      04-08   
127                                 Number Three School      KG-03   
2206                            The Gray Charter School      KG-08   
1829                            Beatrice Gilmore School      03-04   
1830                             Memorial Middle School      05-08   
2211             Jersey City Golden Door Charter School      PK-08   
109                             Sixteenth Avenue School      PK-05   
108                               Gilbert Avenue School      PK-05   
107                               Gantner Avenue School      KG-05   
106                              Memorial Middle School      06-08   
105                         Memorial Senior High School      09-12   
264                                   Midland School #1      PK-08   
208    NORTHERN VALLEY REGIONAL HIGH SCHOOL AT DEMAREST      09-12   
209   Northern Valley Regional High School at Old 

In [633]:
slow_internet_schools = merge_schools_tech[(merge_schools_tech["InternetSpeed"] <= 100) | (merge_schools_tech["InternetSpeed"].isnull())  ]

In [634]:
len(slow_internet_schools.index)

442

In [635]:
math_scores = performance_report['MathParticipationPerform']
ela_scores = performance_report['ELALiteracyParticipationPerform']

In [636]:
school_math_scores = math_scores[(math_scores["StudentGroup"] == "Schoolwide") ]
school_ela_scores = ela_scores[(ela_scores["StudentGroup"] == "Schoolwide") ]

In [637]:
print len(school_math_scores.index)

2239


In [638]:
print len(school_ela_scores.index)

2241


In [639]:
list(school_math_scores.columns)

[u'CountyCode',
 u'DistrictCode',
 u'SchoolCode',
 u'StudentGroup',
 u'Subject',
 u'ValidScores',
 u'ParticipationPercent',
 u'SchoolPerformance',
 u'DistrictPerformance',
 u'StatePerformance',
 u'ProfRateFederalAccountability',
 u'AnnualTarget',
 u'MetTarget']

In [640]:
slow_internet_schools1 =  pd.merge(slow_internet_schools, school_math_scores, left_on=['County Code', 'District Code', 'School Code'], right_on=['CountyCode', 'DistrictCode', 'SchoolCode'], how='left')

In [641]:
slow_internet_schools2 =  pd.merge(slow_internet_schools1, school_ela_scores, left_on=['County Code', 'District Code', 'School Code'], right_on=['CountyCode', 'DistrictCode', 'SchoolCode'], suffixes=('_Math', '_ELA'), how='left')

In [642]:
slow_internet_schools2.head(2)

County Code  District Code  School Code County Name         District Name  \
0            3           4870           50      BERGEN  SOUTH HACKENSACK TWP   
1           37           1980           50      SUSSEX           HAMPTON TWP   

                           School Name Grade Span StudentDeviceRation  \
0                      Memorial School      PK-08               1:0.7   
1  MARIAN E. MCKEOWN ELEMENTARY SCHOOL      KG-06               1:0.8   

   InternetSpeed RecommendedInternetSpeed      ...      StudentGroup_ELA  \
0            3.7                 100 kbps      ...            Schoolwide   
1            5.2                 100 kbps      ...            Schoolwide   

  Subject_ELA ValidScores_ELA ParticipationPercent_ELA  SchoolPerformance_ELA  \
0         LAL             149                     93.1                  73.10   
1         LAL             155                     97.0                  68.40   

   DistrictPerformance_ELA  StatePerformance_ELA  \
0                    73.10                  54.9   
1                    68.40                  54.9   

  ProfRateFederalAccountability_ELA AnnualTarget_ELA MetTarget_ELA  
0                              71.3             65.8    Met Target  
1                              68.4             72.8   Met Target†  

[2 rows x 40 columns]

In [643]:
race_data = performance_report['EnrollmentByRacialEthnicGroup']

In [644]:
race_data.head(10)

CountyCode  DistrictCode  SchoolCode                 RacialAndEthnicGroup  \
0           1            10          50                                Asian   
1           1            10          50            Black or African American   
2           1            10          50                             Hispanic   
3           1            10          50                    Two or More Races   
4           1            10          50                                White   
5           1            10          60                                Asian   
6           1            10          60            Black or African American   
7           1            10          60                             Hispanic   
8           1            10          60  Native Hawaiian or Pacific Islander   
9           1            10          60                    Two or More Races   

   Percent  
0      0.8  
1     16.7  
2     10.3  
3      9.2  
4     63.1  
5      7.2  
6     14.4  
7     14.0  
8      0.4  
9      5.1

In [645]:
asian_data = race_data[(race_data["RacialAndEthnicGroup"] == "Asian") ]
blk_data = race_data[(race_data["RacialAndEthnicGroup"] == "Black or African American") ]
hisp_data = race_data[(race_data["RacialAndEthnicGroup"] == "Hispanic") ]
white_data = race_data[(race_data["RacialAndEthnicGroup"] == "White") ]

In [646]:
slow_internet_schools_wwhite =  pd.merge(slow_internet_schools2, white_data, left_on=['County Code', 'District Code', 'School Code'], right_on=['CountyCode', 'DistrictCode', 'SchoolCode'], how='left', suffixes=('', '_white'))
slow_internet_schools_wblck =  pd.merge(slow_internet_schools_wwhite, blk_data, left_on=['County Code', 'District Code', 'School Code'], right_on=['CountyCode', 'DistrictCode', 'SchoolCode'], how='left', suffixes=('', '_blck'))
slow_internet_schools_whisp =  pd.merge(slow_internet_schools_wblck, hisp_data, left_on=['County Code', 'District Code', 'School Code'], right_on=['CountyCode', 'DistrictCode', 'SchoolCode'], how='left', suffixes=('', '_hisp'))
slow_internet_schools_wrace =  pd.merge(slow_internet_schools_whisp, asian_data, left_on=['County Code', 'District Code', 'School Code'], right_on=['CountyCode', 'DistrictCode', 'SchoolCode'], how='left', suffixes=('_x', '_asian'))


In [647]:
len(slow_internet_schools2.index)

442

In [648]:
len(slow_internet_schools_wwhite.index)

442

In [649]:
len(slow_internet_schools_wrace.index)

442

In [650]:
cols_to_drop = ['CountyCode_Math', 'DistrictCode_Math', 'SchoolCode_Math', 'StudentGroup_Math', 'Subject_Math', 'CountyCode_ELA', 'DistrictCode_ELA', 'SchoolCode_ELA', 'StudentGroup_ELA', 'Subject_ELA', 'CountyCode_x', 'DistrictCode_x', 'SchoolCode_x','CountyCode_blck', 'DistrictCode_blck', 'SchoolCode_blck','CountyCode_hisp', 'DistrictCode_hisp', 'SchoolCode_hisp','CountyCode_asian', 'DistrictCode_asian', 'SchoolCode_asian']
slow_internet_schools_wrace2 = slow_internet_schools_wrace.drop(cols_to_drop, axis=1)

In [651]:
len(slow_internet_schools_wrace2.columns)

38

In [652]:
slow_internet_schools_wrace2.rename(columns={'RacialAndEthnicGroup_x': 'RacialAndEthnicGroup_white', 'Percent_x': 'Percent_white'}, inplace=True)

In [653]:
slow_internet_schools_wrace2.head(5)

County Code  District Code  School Code County Name         District Name  \
0            3           4870           50      BERGEN  SOUTH HACKENSACK TWP   
1           37           1980           50      SUSSEX           HAMPTON TWP   
2            3           1470           60      BERGEN         FAIRVIEW BORO   
3            3           1470           80      BERGEN         FAIRVIEW BORO   
4           80           6665          930    CHARTERS               Gray CS   

                           School Name Grade Span StudentDeviceRation  \
0                      Memorial School      PK-08               1:0.7   
1  MARIAN E. MCKEOWN ELEMENTARY SCHOOL      KG-06               1:0.8   
2                       Lincoln School      04-08               1:0.4   
3                  Number Three School      KG-03               1:0.2   
4              The Gray Charter School      KG-08               1:0.3   

   InternetSpeed RecommendedInternetSpeed      ...      AnnualTarget_ELA  \
0            3.7                 100 kbps      ...                  65.8   
1            5.2                 100 kbps      ...                  72.8   
2            7.0                 100 kbps      ...                  39.6   
3            7.0                 100 kbps      ...                  36.2   
4            9.1                 100 kbps      ...                  67.9   

  MetTarget_ELA RacialAndEthnicGroup_white Percent_white  \
0    Met Target                      White          27.7   
1   Met Target†                      White          93.8   
2    Met Target                      White          17.0   
3   Met Target†                      White          12.1   
4   Met Target†                      White           6.1   

   RacialAndEthnicGroup_blck Percent_blck RacialAndEthnicGroup_hisp  \
0  Black or African American          6.3                  Hispanic   
1  Black or African American          1.0                  Hispanic   
2  Black or African American          1.7                  Hispanic   
3  Black or African American          1.3                  Hispanic   
4  Black or African American         39.4                  Hispanic   

  Percent_hisp  RacialAndEthnicGroup_asian Percent_asian  
0         57.6                       Asian           6.6  
1          4.1                       Asian           0.7  
2         79.4                       Asian           1.7  
3         85.8                       Asian           0.6  
4         50.3                       Asian           3.3  

[5 rows x 38 columns]

In [654]:
cols_to_drop = ['RacialAndEthnicGroup_blck', 'RacialAndEthnicGroup_white', 'RacialAndEthnicGroup_hisp', 'RacialAndEthnicGroup_asian']
slow_internet_schools_wrace3 = slow_internet_schools_wrace2.drop(cols_to_drop, axis=1)

In [655]:
list(slow_internet_schools_wrace3.columns)

[u'County Code',
 u'District Code',
 u'School Code',
 u'County Name',
 u'District Name',
 u'School Name',
 u'Grade Span',
 u'StudentDeviceRation',
 u'InternetSpeed',
 u'RecommendedInternetSpeed',
 u'MetRecommendedSpeed',
 u'ConnectivityBetweenSchools',
 u'RecommendedConnectivity',
 u'MetRecommendedConnectivity',
 u'ValidScores_Math',
 u'ParticipationPercent_Math',
 u'SchoolPerformance_Math',
 u'DistrictPerformance_Math',
 u'StatePerformance_Math',
 u'ProfRateFederalAccountability_Math',
 u'AnnualTarget_Math',
 u'MetTarget_Math',
 u'ValidScores_ELA',
 u'ParticipationPercent_ELA',
 u'SchoolPerformance_ELA',
 u'DistrictPerformance_ELA',
 u'StatePerformance_ELA',
 u'ProfRateFederalAccountability_ELA',
 u'AnnualTarget_ELA',
 u'MetTarget_ELA',
 'Percent_white',
 u'Percent_blck',
 u'Percent_hisp',
 u'Percent_asian']

In [656]:
slow_internet_schools_wrace3

County Code  District Code  School Code County Name  \
0              3           4870           50      BERGEN   
1             37           1980           50      SUSSEX   
2              3           1470           60      BERGEN   
3              3           1470           80      BERGEN   
4             80           6665          930    CHARTERS   
5             31           5690           60     PASSAIC   
6             31           5690           70     PASSAIC   
7             80           6915          950    CHARTERS   
8              3           1345           90      BERGEN   
9              3           1345           80      BERGEN   
10             3           1345           70      BERGEN   
11             3           1345           60      BERGEN   
12             3           1345           50      BERGEN   
13             3           4470           50      BERGEN   
14             3           3710           50      BERGEN   
15             3           3710           60      BERGEN   
16             5           5890           30  BURLINGTON   
17            31           4010          310     PASSAIC   
18            31           4010           20     PASSAIC   
19            31           4010          309     PASSAIC   
20            31           4010          311     PASSAIC   
21            31           4010          306     PASSAIC   
22            31           4010          308     PASSAIC   
23            31           4010           43     PASSAIC   
24            31           4010          305     PASSAIC   
25            31           4010          325     PASSAIC   
26            31           4010          316     PASSAIC   
27            31           4010           47     PASSAIC   
28            31           4010          315     PASSAIC   
29            31           4010           50     PASSAIC   
..           ...            ...          ...         ...   
412           31            900          100     PASSAIC   
413           31            900          200     PASSAIC   
414           31            900          120     PASSAIC   
415           31            900          110     PASSAIC   
416           80           6020          985    CHARTERS   
417           37           5360           25      SUSSEX   
418           37           5360           20      SUSSEX   
419           37           5360           23      SUSSEX   
420           37           5360           40      SUSSEX   
421           37           5360           30      SUSSEX   
422           31           1920           15     PASSAIC   
423           23           3530          300   MIDDLESEX   
424           23           3530          125   MIDDLESEX   
425           23           3530          123   MIDDLESEX   
426           23           3530          100   MIDDLESEX   
427           23           3530           90   MIDDLESEX   
428           23           3530           80   MIDDLESEX   
429           23           3530           60   MIDDLESEX   
430           23           3530           55   MIDDLESEX   
431           23           3530           50   MIDDLESEX   
432           23           3530          140   MIDDLESEX   
433           23           3530          110   MIDDLESEX   
434           80           6017          932    CHARTERS   
435            7           1799          111      CAMDEN   
436            7           1801          113      CAMDEN   
437           80           6060          942    CHARTERS   
438           80           6101          992    CHARTERS   
439           80           6104          995    CHARTERS   
440           80           6107          998    CHARTERS   
441           80           6183          914    CHARTERS   

                                         District Name  \
0                                 SOUTH HACKENSACK TWP   
1                                          HAMPTON TWP   
2                                        FAIRVIEW BORO   
3                                        FAIRVIEW BO

In [657]:
enr_trends_bystudentgroup = performance_report['EnrollmentTrendsByStudentGroup']

In [658]:
enr_trends_bystudentgroup.head(5)

CountyCode  DistrictCode  SchoolCode       StudentGroup  Percent
0           1            10          50         Disability      9.2
1           1            10          50            Econdis     36.7
2           1            10          50             Female     48.9
3           1            10          50  Homeless Students      0.6
4           1            10          50                LEP      2.2

In [659]:
school_econdis = enr_trends_bystudentgroup[(enr_trends_bystudentgroup["StudentGroup"] == "Econdis") ]
slow_internet_wecondis =  pd.merge(slow_internet_schools_wrace3, school_econdis, left_on=['County Code', 'District Code', 'School Code'], right_on=['CountyCode', 'DistrictCode', 'SchoolCode'], how="left")

In [660]:
len(school_econdis.index)

2453

In [661]:
slow_internet_wecondis.head(5)

County Code  District Code  School Code County Name         District Name  \
0            3           4870           50      BERGEN  SOUTH HACKENSACK TWP   
1           37           1980           50      SUSSEX           HAMPTON TWP   
2            3           1470           60      BERGEN         FAIRVIEW BORO   
3            3           1470           80      BERGEN         FAIRVIEW BORO   
4           80           6665          930    CHARTERS               Gray CS   

                           School Name Grade Span StudentDeviceRation  \
0                      Memorial School      PK-08               1:0.7   
1  MARIAN E. MCKEOWN ELEMENTARY SCHOOL      KG-06               1:0.8   
2                       Lincoln School      04-08               1:0.4   
3                  Number Three School      KG-03               1:0.2   
4              The Gray Charter School      KG-08               1:0.3   

   InternetSpeed RecommendedInternetSpeed   ...   MetTarget_ELA Percent_white  \
0            3.7                 100 kbps   ...      Met Target          27.7   
1            5.2                 100 kbps   ...     Met Target†          93.8   
2            7.0                 100 kbps   ...      Met Target          17.0   
3            7.0                 100 kbps   ...     Met Target†          12.1   
4            9.1                 100 kbps   ...     Met Target†           6.1   

  Percent_blck Percent_hisp Percent_asian CountyCode DistrictCode SchoolCode  \
0          6.3         57.6           6.6        3.0       4870.0       50.0   
1          1.0          4.1           0.7       37.0       1980.0       50.0   
2          1.7         79.4           1.7        3.0       1470.0       60.0   
3          1.3         85.8           0.6        3.0       1470.0       80.0   
4         39.4         50.3           3.3       80.0       6665.0      930.0   

   StudentGroup Percent  
0       Econdis    38.4  
1       Econdis    14.8  
2       Econdis    72.2  
3       Econdis    81.8  
4       Econdis    78.8  

[5 rows x 39 columns]

In [662]:
cols_to_drop = ['CountyCode', 'DistrictCode', 'SchoolCode', 'StudentGroup']
slow_internet_schools_all_joins = slow_internet_wecondis.drop(cols_to_drop, axis=1)

In [663]:
slow_internet_schools_all_joins.rename(columns={'Percent': 'Percent_econdis'}, inplace=True)

In [664]:
slow_internet_schools_all_joins.head(5)

County Code  District Code  School Code County Name         District Name  \
0            3           4870           50      BERGEN  SOUTH HACKENSACK TWP   
1           37           1980           50      SUSSEX           HAMPTON TWP   
2            3           1470           60      BERGEN         FAIRVIEW BORO   
3            3           1470           80      BERGEN         FAIRVIEW BORO   
4           80           6665          930    CHARTERS               Gray CS   

                           School Name Grade Span StudentDeviceRation  \
0                      Memorial School      PK-08               1:0.7   
1  MARIAN E. MCKEOWN ELEMENTARY SCHOOL      KG-06               1:0.8   
2                       Lincoln School      04-08               1:0.4   
3                  Number Three School      KG-03               1:0.2   
4              The Gray Charter School      KG-08               1:0.3   

   InternetSpeed RecommendedInternetSpeed       ...        \
0            3.7                 100 kbps       ...         
1            5.2                 100 kbps       ...         
2            7.0                 100 kbps       ...         
3            7.0                 100 kbps       ...         
4            9.1                 100 kbps       ...         

  DistrictPerformance_ELA StatePerformance_ELA  \
0                   73.10                 54.9   
1                   68.40                 54.9   
2                   40.70                 54.9   
3                   40.70                 54.9   
4                   77.60                 54.9   

  ProfRateFederalAccountability_ELA AnnualTarget_ELA MetTarget_ELA  \
0                              71.3             65.8    Met Target   
1                              68.4             72.8   Met Target†   
2                              42.6             39.6    Met Target   
3                              30.5             36.2   Met Target†   
4                              67.4             67.9   Met Target†   

  Percent_white Percent_blck Percent_hisp  Percent_asian Percent_econdis  
0          27.7          6.3         57.6            6.6            38.4  
1          93.8          1.0          4.1            0.7            14.8  
2          17.0          1.7         79.4            1.7            72.2  
3          12.1          1.3         85.8            0.6            81.8  
4           6.1         39.4         50.3            3.3            78.8  

[5 rows x 35 columns]

<a id='analysis'></a>

In [665]:
slow_internet_schools_all_joins.to_csv('results/slow_internet_schools.csv', index=False, encoding='utf-8')

# Analysis

In [666]:
slow_schools = len(slow_internet_schools_all_joins.index)

In [667]:
recommended_speed

MetRecommendedSpeed  School Code    Percent
0                  No          442  19.767442
1                 Yes         1794  80.232558

According to the data, about **19.8%** or **442** of New Jersey schools are labeled as not meeting the state recommended Internet speed of 100 kbps– meaning they have internet speeds of less than 100 kbps. **Note:** there are about 7 schools which have their actual Internet speed as null but NJ DOE has labeled them as not having met the recommended Internet Speed. 

This is based on data on the **2,236** New Jersey schools that reported tech readiness results. 

### 25 Schools With Slowest Internet Speeds 

In [668]:
merge_schools_tech.head(25)

County Code  District Code  School Code County Name  \
275             3           4870           50      BERGEN   
1939           37           1980           50      SUSSEX   
126             3           1470           60      BERGEN   
127             3           1470           80      BERGEN   
2206           80           6665          930    CHARTERS   
1829           31           5690           60     PASSAIC   
1830           31           5690           70     PASSAIC   
2211           80           6915          950    CHARTERS   
109             3           1345           90      BERGEN   
108             3           1345           80      BERGEN   
107             3           1345           70      BERGEN   
106             3           1345           60      BERGEN   
105             3           1345           50      BERGEN   
264             3           4470           50      BERGEN   
208             3           3710           50      BERGEN   
209             3           3710           60      BERGEN   
421             5           5890           30  BURLINGTON   
1788           31           4010          310     PASSAIC   
1752           31           4010           20     PASSAIC   
1787           31           4010          309     PASSAIC   
1789           31           4010          311     PASSAIC   
1784           31           4010          306     PASSAIC   
1786           31           4010          308     PASSAIC   
1754           31           4010           43     PASSAIC   
1783           31           4010          305     PASSAIC   

                               District Name  \
275                     SOUTH HACKENSACK TWP   
1939                             HAMPTON TWP   
126                            FAIRVIEW BORO   
127                            FAIRVIEW BORO   
2206                                 Gray CS   
1829                           WOODLAND PARK   
1830                           WOODLAND PARK   
2211  Jersey City Golden Door Charter School   
109                             ELMWOOD PARK   
108                             ELMWOOD PARK   
107                             ELMWOOD PARK   
106                             ELMWOOD PARK   
105                             ELMWOOD PARK   
264                        ROCHELLE PARK TWP   
208                 NORTHERN VALLEY REGIONAL   
209                 NORTHERN VALLEY REGIONAL   
421                             WOODLAND TWP   
1788                           PATERSON CITY   
1752                           PATERSON CITY   
1787                           PATERSON CITY   
1789                           PATERSON CITY   
1784                           PATERSON CITY   
1786                           PATERSON CITY   
1754                           PATERSON CITY   
1783                           PATERSON CITY   

                                            School Name Grade Span  \
275                                     Memorial School      PK-08   
1939                MARIAN E. MCKEOWN ELEMENTARY SCHOOL      KG-06   
126                                      Lincoln School      04-08   
127                                 Number Three School      KG-03   
2206                            The Gray Charter School      KG-08   
1829                            Beatrice Gilmore School      03-04   
1830                             Memorial Middle School      05-08   
2211             Jersey City Golden Door Charter School      PK-08   
109                             Sixteenth Avenue School      PK-05   
108                               Gilbert Avenue School      PK-05   
107                               Gantner Avenue School      KG-05   
106                              Memorial Middle School      06-08   
105                         Memorial Senior High School      09-12   
264                                   Midland School #1      PK-08   
208    NORTHERN VALLEY REGIONAL HIGH SCHOOL AT DEMAREST      09-12   
209   Northern Valley Regional High School at Old 

## How did these Schools Score in Math?

In [669]:
slow_internet_schools_all_joins['MetTarget_Math'].unique()

array([u'Met Target', u'Met Target\u2020', u'Not Met', u'N', u'Met Goal',
       u'**', nan], dtype=object)

In [670]:
met_target_math = slow_internet_schools_all_joins.groupby( [ "MetTarget_Math"] ).count()

In [671]:
met_target_math = met_target_math[['School Code']]
met_target_math.reset_index()
met_target_math['Percent'] = (met_target_math['School Code']/slow_schools)*100
met_target_math.reset_index()

MetTarget_Math  School Code    Percent
0             **            2   0.452489
1       Met Goal            5   1.131222
2     Met Target          165  37.330317
3    Met Target†          126  28.506787
4              N           12   2.714932
5        Not Met          131  29.638009

Out all of these 442 schools with less than 100 kpbs internet speeds, about **29.6%** or **131 schools** of them did not meet their target PARCC Math scores, according to NJ DOE.

**291 schools** or about **65.8%** of schools either met their PARCC Math score targets or met their PARCC Math score targets within a confidence interval. 

**Note:** An "*" indicates that data is not displayed to protect student privacy; "N" indicates no data is available to display

## How did these Schools Score in Language Arts?

In [672]:
met_target_ela = slow_internet_schools_all_joins.groupby( [ "MetTarget_ELA"] ).count()
met_target_ela = met_target_ela[['School Code']]
met_target_ela.reset_index()
met_target_ela['Percent'] = (met_target_ela['School Code']/slow_schools)*100
met_target_ela.reset_index()

MetTarget_ELA  School Code    Percent
0            **            2   0.452489
1      Met Goal           15   3.393665
2    Met Target          225  50.904977
3   Met Target†          116  26.244344
4             N           10   2.262443
5       Not Met           73  16.515837

Out all of these 442 schools with less than 100 kpbs internet speeds, about **16.5%** or **73 schools** of them did not meet their target PARCC English Langauge Arts scores, according to NJ DOE.

**341 schools** or about **77.3%** of schools either met their PARCC ELA score targets or met their PARCC ELA score targets within a confidence interval. 

**Note:** An "*" indicates that data is not displayed to protect student privacy; "N" indicates no data is available to display

## Racial Breakdown

### How many are majority minority (less than 50 percent white)? 

In [673]:
percent_majority_minority = slow_internet_schools_all_joins[slow_internet_schools_all_joins['Percent_white'] <= 50]

In [674]:
num_majority_minority = len(percent_majority_minority.index)
num_majority_minority

277

In [675]:
pct_majority_minority = (float(num_majority_minority)/float(slow_schools))*100
pct_majority_minority

62.66968325791855

About **62.7%** or **277** schools are majority minority, as in more than 50 percent of the students are non-white. 

### How many are majority black (more than 50 percent black)? 

In [676]:
percent_majority_blck = slow_internet_schools_all_joins[slow_internet_schools_all_joins['Percent_blck'] > 50]

In [677]:
num_majority_blck = len(percent_majority_blck.index)
num_majority_blck

94

In [678]:
pct_majority_blck = (float(num_majority_blck)/float(slow_schools))*100
pct_majority_blck

21.266968325791854

About **21.3%** or **94** schools are majority Black, as in more than 50 percent of the students are Black. 

#### How many are majority Hispanic (more than 50 percent Hispanic/Latino origin)?

In [679]:
percent_majority_hisp = slow_internet_schools_all_joins[slow_internet_schools_all_joins['Percent_hisp'] > 50]

In [680]:
num_majority_hisp = len(percent_majority_hisp.index)
num_majority_hisp

135

In [681]:
pct_majority_hisp = (float(num_majority_hisp)/float(slow_schools))*100
pct_majority_hisp

30.542986425339368

About **30.5%** or **135** schools are majority Hispanic, as in more than 50 percent of the students are Hispanic. 

#### How many are majority Asian (more than 50 percent Asian)?

In [682]:
percent_majority_asian = slow_internet_schools_all_joins[slow_internet_schools_all_joins['Percent_asian'] > 50]

In [683]:
num_majority_asian = len(percent_majority_asian.index)
num_majority_asian

0

None of the schools are more than 50 percent Asian.

## Economic Disadvantage Breakdown

#### How many schools have a majority of students at an economic disadvantage (more than 50 percent)?

In [684]:
percent_majority_econdis = slow_internet_schools_all_joins[slow_internet_schools_all_joins['Percent_econdis'] > 50]

In [685]:
num_majority_econdis = len(percent_majority_econdis.index)
num_majority_econdis

254

In [686]:
pct_majority_econdis = (float(num_majority_econdis)/float(slow_schools))*100
pct_majority_econdis

57.466063348416284

About **57.5** or **254** schools have the majority of their students at an economic disadvantage, as in more than 50 percent of the students are Hispanic. 